File: packages/js-sdk/src/modules/webhooks.ts
/**
 * TurboWebhooks Module - Org-scoped webhook subscription management
 *
 * Wraps the backend `/api/webhooks/*` surface. All routes require an admin
 * `TDX-` API key. Webhook management does not send signature emails, so
 * `skipSenderValidation: true` is hardcoded inside `configure()` and
 * `getClient()` to avoid the senderEmail-required `ValidationError` from
 * `HttpClient`.
 */

import { HttpClient, HttpClientConfig } from '../http';
import {
  CreateWebhookRequest,
  CreateWebhookResponse,
  ListDeliveriesRequest,
  ListDeliveriesResponse,
  ListWebhooksRequest,
  ListWebhooksResponse,
  NotifyWebhookResponse,
  RegenerateSecretResponse,
  ReplayDeliveryResponse,
  TestWebhookRequest,
  TestWebhookResponse,
  UpdateWebhookRequest,
  Webhook,
  WebhookStats,
  WebhookStatsRequest,
  WebhookWithStats,
} from '../types/webhooks';

export class TurboWebhooks {
  private static client: HttpClient;

  /**
   * Configure the TurboWebhooks module with API credentials.
   *
   * Mirrors `TurboPartner.configure()` — extracts only the fields needed
   * and hardcodes `skipSenderValidation: true` as a literal so a caller
   * cannot accidentally override it by passing `skipSenderValidation: false`.
   *
   * @param config - Configuration object
   * @param config.apiKey - TurboDocx API key (required, must be administrator role)
   * @param config.orgId - Organization ID (required)
   * @param config.baseUrl - API base URL (optional, defaults to https://api.turbodocx.com)
   *
   * @example
   * ```typescript
   * TurboWebhooks.configure({
   *   apiKey: process.env.TURBODOCX_API_KEY,
   *   orgId: process.env.TURBODOCX_ORG_ID,
   * });
   * ```
   */
  static configure(config: HttpClientConfig): void {
    this.client = new HttpClient({
      apiKey: config.apiKey,
      accessToken: config.accessToken,
      orgId: config.orgId,
      baseUrl: config.baseUrl,
      skipSenderValidation: true,
    });
  }

  /**
   * Get the HTTP client instance, initializing from env vars if necessary.
   * Mirrors `TurboPartner.getClient()` — fails loudly with a descriptive
   * error rather than silently auto-configuring.
   */
  private static getClient(): HttpClient {
    if (!this.client) {
      const apiKey = process.env.TURBODOCX_API_KEY;
      const orgId = process.env.TURBODOCX_ORG_ID;
      if (!apiKey || !orgId) {
        throw new Error(
          'TurboWebhooks must be configured before use. Call TurboWebhooks.configure() ' +
            'or set TURBODOCX_API_KEY and TURBODOCX_ORG_ID environment variables.',
        );
      }
      this.configure({ apiKey, orgId });
    }
    return this.client;
  }

  /**
   * Create a webhook subscription. The returned `secret` is shown ONCE and
   * must be saved by the caller; it is never returned again by any other
   * endpoint. Webhook URLs must use HTTPS.
   */
  static async createWebhook(input: CreateWebhookRequest): Promise<CreateWebhookResponse> {
    // Backend response is `{ data: { id, secret }, message }`. Because `data`
    // is not the sole top-level key, the HttpClient's smartUnwrap returns the
    // envelope as-is. Pull `.data` here so the SDK return shape stays domain-focused.
    const envelope = await this.getClient().post<{ data: CreateWebhookResponse }>(
      '/api/webhooks',
      input,
    );
    return envelope.data;
  }

  /** List webhook subscriptions for the configured org. */
  static async listWebhooks(input: ListWebhooksRequest = {}): Promise<ListWebhooksResponse> {
    return this.getClient().get<ListWebhooksResponse>('/api/webhooks', input);
  }

  /** Get a single webhook by name with current delivery stats and available events. */
  static async getWebhook(name: string): Promise<WebhookWithStats> {
    return this.getClient().get<WebhookWithStats>(
      `/api/webhooks/${encodeURIComponent(name)}`,
    );
  }

  /** Patch one or more fields on an existing webhook. */
  static async updateWebhook(name: string, patch: UpdateWebhookRequest): Promise<Webhook> {
    // Backend returns `{ data: webhook, message }` — same envelope quirk as
    // createWebhook. smartUnwrap leaves it alone because of the extra `message`
    // key, so we extract `.data` explicitly.
    const envelope = await this.getClient().patch<{ data: Webhook }>(
      `/api/webhooks/${encodeURIComponent(name)}`,
      patch,
    );
    return envelope.data;
  }

  /** Soft-delete a webhook and its delivery history. */
  static async deleteWebhook(name: string): Promise<{ message: string }> {
    return this.getClient().delete<{ message: string }>(
      `/api/webhooks/${encodeURIComponent(name)}`,
    );
  }

  /** Send a test delivery to all URLs configured on the webhook. */
  static async testWebhook(
    name: string,
    input: TestWebhookRequest,
  ): Promise<TestWebhookResponse> {
    // Envelope: `{ data: { deliveries, summary }, message }`
    const envelope = await this.getClient().post<{ data: TestWebhookResponse }>(
      `/api/webhooks/${encodeURIComponent(name)}/test`,
      input,
    );
    return envelope.data;
  }

  /**
   * Send a manual notification to all URLs configured on the webhook.
   *
   * NOTE: This routes to the same internal delivery path as `testWebhook` and
   * returns the same payload shape. The backend differentiates the two only in
   * the response/error message strings ("Test webhook sent..." vs
   * "Manual notification sent...", "Cannot test inactive webhook" vs
   * "Cannot send notification to inactive webhook"). Prefer `testWebhook` in
   * new code; `notifyWebhook` exists for symmetry with the backend route.
   */
  static async notifyWebhook(
    name: string,
    input: TestWebhookRequest,
  ): Promise<NotifyWebhookResponse> {
    // Envelope: `{ data: { deliveries, summary }, message }`
    const envelope = await this.getClient().post<{ data: NotifyWebhookResponse }>(
      `/api/webhooks/${encodeURIComponent(name)}/notify`,
      input,
    );
    return envelope.data;
  }

  /**
   * Rotate the webhook's HMAC secret. The new secret is shown ONCE in the
   * response and must be saved; old signatures will fail immediately.
   */
  static async regenerateWebhookSecret(name: string): Promise<RegenerateSecretResponse> {
    // Envelope: `{ data: { id, secret, regeneratedAt }, message }`
    const envelope = await this.getClient().post<{ data: RegenerateSecretResponse }>(
      `/api/webhooks/${encodeURIComponent(name)}/regenerate`,
    );
    return envelope.data;
  }

  /** List historical delivery attempts for a webhook, with optional filters. */
  static async listWebhookDeliveries(
    name: string,
    input: ListDeliveriesRequest = {},
  ): Promise<ListDeliveriesResponse> {
    return this.getClient().get<ListDeliveriesResponse>(
      `/api/webhooks/${encodeURIComponent(name)}/deliveries`,
      input,
    );
  }

  /** Manually retry a specific past delivery by ID. */
  static async replayWebhookDelivery(
    name: string,
    deliveryId: string,
  ): Promise<ReplayDeliveryResponse> {
    // Envelope: `{ data: { id, httpStatus }, message }`
    const envelope = await this.getClient().post<{ data: ReplayDeliveryResponse }>(
      `/api/webhooks/${encodeURIComponent(name)}/replay`,
      { deliveryId },
    );
    return envelope.data;
  }

  /** Aggregate delivery stats for the webhook over a sliding window (days). */
  static async getWebhookStats(
    name: string,
    input: WebhookStatsRequest = {},
  ): Promise<WebhookStats> {
    return this.getClient().get<WebhookStats>(
      `/api/webhooks/${encodeURIComponent(name)}/stats`,
      input,
    );
  }
}

File: packages/py-sdk/src/turbodocx_sdk/modules/webhooks.py
"""
TurboWebhooks Module - Org-scoped signature webhook subscription

The SDK is intentionally locked to a single webhook per org, identified by
the fixed name `signature`. This matches the UI's Signature Webhooks
settings page so SDK-managed and UI-managed webhooks stay in sync. To
manage multiple webhooks per org, call the REST API directly.

All routes require an admin TDX- API key. Webhook management does not
send signature emails, so HttpClient is constructed with
skip_sender_validation=True.

POST/PATCH responses come back as `{"data": ..., "message": ...}` envelopes
which the HttpClient's _smart_unwrap leaves intact (it only unwraps
single-key `{"data": ...}` responses). Methods that hit non-GET routes
extract `.data` explicitly. GET routes are auto-unwrapped.
"""

import os
from typing import Any, Dict, List, Optional
from urllib.parse import urlencode

from ..http import HttpClient, AuthenticationError


SIGNATURE_WEBHOOK_NAME = "signature"


def _build_query_string(params: Dict[str, Any]) -> str:
    """Build URL query string from non-None parameters. Booleans lowercased."""
    filtered = {k: v for k, v in params.items() if v is not None}
    if not filtered:
        return ""
    for k, v in filtered.items():
        if isinstance(v, bool):
            filtered[k] = str(v).lower()
    return "?" + urlencode(filtered)


class TurboWebhooks:
    """TurboWebhooks module for managing the org's signature webhook."""

    _client: Optional[HttpClient] = None

    @classmethod
    def configure(
        cls,
        api_key: Optional[str] = None,
        org_id: Optional[str] = None,
        base_url: str = "https://api.turbodocx.com",
    ) -> None:
        """
        Configure the TurboWebhooks module with API credentials.

        Args:
            api_key: TurboDocx API key (must be administrator role)
            org_id: Organization ID
            base_url: API base URL (defaults to https://api.turbodocx.com)

        Example:
            >>> TurboWebhooks.configure(
            ...     api_key=os.environ["TURBODOCX_API_KEY"],
            ...     org_id=os.environ["TURBODOCX_ORG_ID"],
            ... )
        """
        client = HttpClient(
            api_key=api_key,
            org_id=org_id,
            base_url=base_url,
            skip_sender_validation=True,
        )
        # Webhook routes are org-scoped and admin-gated — org_id is required
        # (HttpClient leaves it optional for TurboSign).
        if client.org_id is None:
            raise AuthenticationError(
                "Organization ID (org_id) is required for authentication"
            )
        cls._client = client

    @classmethod
    def _get_client(cls) -> HttpClient:
        """
        Get the HTTP client. Auto-configures from env vars if not yet configured.
        Mirrors TurboPartner's behavior: explicit env-var check with a clear
        error rather than silent auto-configure.
        """
        if cls._client is None:
            api_key = os.environ.get("TURBODOCX_API_KEY")
            org_id = os.environ.get("TURBODOCX_ORG_ID")
            if not api_key or not org_id:
                raise RuntimeError(
                    "TurboWebhooks not configured. Call TurboWebhooks.configure("
                    "api_key='...', org_id='...') first, or set "
                    "TURBODOCX_API_KEY and TURBODOCX_ORG_ID environment variables."
                )
            cls.configure(api_key=api_key, org_id=org_id)
        return cls._client  # type: ignore[return-value]

    # ============================================
    # CRUD - always hits /api/webhooks/signature[/...]
    # ============================================

    @classmethod
    async def create_webhook(
        cls,
        urls: List[str],
        events: List[str],
    ) -> Dict[str, Any]:
        """
        Create the org's signature webhook.

        Args:
            urls: List of HTTPS URLs to deliver events to (HTTP returns 400).
                Min 1, max 10.
            events: List of event types. Min 1. Prefer the exported constants
                (WEBHOOK_EVENT_COMPLETED, WEBHOOK_EVENT_RECIPIENT_SIGNED, ...)
                over raw strings; WEBHOOK_EVENTS holds all 7. Plain strings are
                still accepted so new backend events work without an SDK bump.

        Returns:
            Dict with `id` and `secret`. The `secret` is shown ONCE; store it
            immediately on receipt.

        Raises:
            ValidationError: if URLs are not HTTPS (HTTP 400)
            ConflictError: if a webhook with this name already exists (HTTP 409)
            AuthorizationError: if the API key lacks admin role
        """
        envelope = await cls._get_client().post(
            "/api/webhooks",
            data={"name": SIGNATURE_WEBHOOK_NAME, "urls": urls, "events": events},
        )
        return envelope["data"]

    @classmethod
    async def get_webhook(cls) -> Dict[str, Any]:
        """
        Get the org's signature webhook with delivery stats + the server-
        provided list of subscribable events.
        """
        return await cls._get_client().get(
            f"/api/webhooks/{SIGNATURE_WEBHOOK_NAME}"
        )

    @classmethod
    async def update_webhook(
        cls,
        *,
        urls: Optional[List[str]] = None,
        events: Optional[List[str]] = None,
        is_active: Optional[bool] = None,
    ) -> Dict[str, Any]:
        """
        Patch one or more fields on the signature webhook.

        Renaming is not supported — the SDK manages a fixed name.

        `urls` and `events` are optional, but they still enforce their minimums
        when present: passing an empty list is a 400. Leave the argument as None
        to leave the field untouched. `urls` also caps at 10 entries.
        """
        body: Dict[str, Any] = {}
        if urls is not None:
            body["urls"] = urls
        if events is not None:
            body["events"] = events
        if is_active is not None:
            body["isActive"] = is_active

        envelope = await cls._get_client().patch(
            f"/api/webhooks/{SIGNATURE_WEBHOOK_NAME}",
            data=body,
        )
        return envelope["data"]

    @classmethod
    async def delete_webhook(cls) -> Dict[str, Any]:
        """Soft-delete the signature webhook and its delivery history."""
        return await cls._get_client().delete(
            f"/api/webhooks/{SIGNATURE_WEBHOOK_NAME}"
        )

    # ============================================
    # TEST / NOTIFY
    # ============================================

    @classmethod
    async def test_webhook(
        cls,
        event_type: str,
        payload: Dict[str, Any],
    ) -> Dict[str, Any]:
        """
        Send a test delivery to all URLs configured on the signature webhook.

        Returns:
            Dict with `deliveries` (list) and `summary` (total/successful/failed).
        """
        envelope = await cls._get_client().post(
            f"/api/webhooks/{SIGNATURE_WEBHOOK_NAME}/test",
            data={"eventType": event_type, "payload": payload},
        )
        return envelope["data"]

    @classmethod
    async def notify_webhook(
        cls,
        event_type: str,
        payload: Dict[str, Any],
    ) -> Dict[str, Any]:
        """
        Send a manual notification to all URLs configured on the signature
        webhook.

        NOTE: routes through the same backend handler as `test_webhook` and
        returns the same shape; the only wire-level difference is the response
        message string. Prefer `test_webhook` in new code.
        """
        envelope = await cls._get_client().post(
            f"/api/webhooks/{SIGNATURE_WEBHOOK_NAME}/notify",
            data={"eventType": event_type, "payload": payload},
        )
        return envelope["data"]

    # ============================================
    # DELIVERIES + REPLAY
    # ============================================

    @classmethod
    async def list_webhook_deliveries(
        cls,
        *,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        event_type: Optional[str] = None,
        is_delivered: Optional[bool] = None,
        http_status: Optional[int] = None,
    ) -> Dict[str, Any]:
        """List historical delivery attempts for the signature webhook."""
        qs = _build_query_string(
            {
                "limit": limit,
                "offset": offset,
                "eventType": event_type,
                "isDelivered": is_delivered,
                "httpStatus": http_status,
            }
        )
        return await cls._get_client().get(
            f"/api/webhooks/{SIGNATURE_WEBHOOK_NAME}/deliveries{qs}"
        )

    @classmethod
    async def replay_webhook_delivery(
        cls,
        delivery_id: str,
    ) -> Dict[str, Any]:
        """Manually retry a specific past delivery by ID."""
        envelope = await cls._get_client().post(
            f"/api/webhooks/{SIGNATURE_WEBHOOK_NAME}/replay",
            data={"deliveryId": delivery_id},
        )
        return envelope["data"]

    # ============================================
    # SECRET ROTATION + STATS
    # ============================================

    @classmethod
    async def regenerate_webhook_secret(cls) -> Dict[str, Any]:
        """
        Rotate the webhook's HMAC secret. The new secret is shown ONCE in the
        response and must be saved; old signatures will fail immediately.
        """
        envelope = await cls._get_client().post(
            f"/api/webhooks/{SIGNATURE_WEBHOOK_NAME}/regenerate",
            data=None,
        )
        return envelope["data"]

    @classmethod
    async def get_webhook_stats(
        cls,
        *,
        days: Optional[int] = None,
    ) -> Dict[str, Any]:
        """Aggregate delivery stats for the webhook over a sliding window."""
        qs = _build_query_string({"days": days})
        return await cls._get_client().get(
            f"/api/webhooks/{SIGNATURE_WEBHOOK_NAME}/stats{qs}"
        )
